main/test/ca/uwaterloo/flix/library/TestIterator.flix

```diff
@@ -475,6 +475,108 @@ namespace TestIterator {
+    // foldLeft                                                                //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def foldLeft01(): Bool & Impure =
+        Nil: List[Int32] |> Iterator.toIter |> Iterator.foldLeft((acc, n) -> n :: acc, Nil: List[Int32]) == Nil: List[Int32]
+
+    @test
+    def foldLeft02(): Bool & Impure =
+        Iterator.range(0, 1000) |> Iterator.foldLeft((acc, n) -> n :: acc, Nil: List[Int32]) ==
+            List.range(0, 1000) |> List.reverse
+
+    @test
+    def foldLeft03(): Bool & Impure =
+        let iter = Iterator.range(0, 1000);
+        iter |> Iterator.foldLeft((_, _) -> true, true);
+        iter |> Iterator.isEmpty
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // foldRight                                                               //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def foldRight01(): Bool & Impure =
+        Nil: List[Int32] |> Iterator.toIter |> Iterator.foldRight((n, acc) -> n :: acc, Nil: List[Int32]) == Nil: List[Int32]
+
+    @test
+    def foldRight02(): Bool & Impure =
+        Iterator.range(0, 1000) |> Iterator.foldRight((n, acc) -> n :: acc, Nil: List[Int32]) ==
+            List.range(0, 1000)
+
+    @test
+    def foldRight03(): Bool & Impure =
+        let iter = Iterator.range(0, 1000);
+        iter |> Iterator.foldRight((_, _) -> true, true);
+        iter |> Iterator.isEmpty
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // takeWhile (pure)                                                        //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def takeWhilePure01(): Bool & Impure =
+        Iterator.repeat(0, 0) |> Iterator.takeWhile(x -> x > 100) |> Iterator.toList == Nil: List[Int32]
+
+    @test
+    def takeWhilePure02(): Bool & Impure =
+        Iterator.range(0, 1000) |> Iterator.takeWhile(x -> x > 100) |> Iterator.toList == Nil: List[Int32]
+
+    @test
+    def takeWhilePure03(): Bool & Impure =
+        Iterator.range(0, 1000) |> Iterator.takeWhile(x -> x < 100) |> Iterator.toList == List.range(0, 100)
+
+    @test
+    def takeWhilePure04(): Bool & Impure =
+        Iterator.repeat(0, 0) |> Iterator.takeWhile(x -> x < 100) |> Iterator.toList == Nil: List[Int32]
+
+    @test
+    def takeWhilePure05(): Bool & Impure =
+        Iterator.repeat(10, 1000) |> Iterator.takeWhile(x -> x < 100) |> Iterator.toList == List.repeat(10, 1000)
+
+    @test
+    def takeWhilePure06(): Bool & Impure =
+        let iter = Iterator.range(0, 1000);
+        iter |> Iterator.takeWhile(x -> x < 100) |> Iterator.toList == List.range(0, 100) and
+            iter |> Iterator.toList == List.range(101, 1000) // Element '100' is consumed by takeWhile
```

Comment:
Should `takeWhile` somehow **not** consume element '100'? It at least has to consume that element to test that the predicate does not hold. The original iterator should then be affected too.

Reply:
Good observation. I don't know the answer. Ill think about it!

Reply:
I guess it is almost impossible to avoid it, since there is no way to "peek" an element.

Reply:
Note Scala has this to say:

![image](https://user-images.githubusercontent.com/2287953/126685693-a6c6ac54-82e0-4d1e-b690-edd7834cf855.png)
![image](https://user-images.githubusercontent.com/2287953/126685786-dcefa4c7-74d4-40dc-9240-10edc8ddae5a.png)

Reply:
So, in short, I think it is fine, and we should consider the original iterator discarded after the take/takeWhile, and the tests should not bother with it.

Reply:
Possibly we should extend this to all operations that transform an iterator.

Reply:
I agree, that would certainly be consistent.